File: scripts/event_identity.py
from __future__ import annotations

import json
import math
import re
import unicodedata
from datetime import date, datetime
from pathlib import Path
from typing import Any, Iterable, Mapping
from urllib.parse import parse_qsl, urlencode, urlparse, urlunparse

ROOT = Path(__file__).resolve().parents[1]
DEFAULT_CONTRACT_PATH = ROOT / "data" / "event_identity_contract.json"


def _text(value: Any) -> str:
    return str(value or "").strip()


def load_event_identity_contract(path: Path | None = None) -> dict[str, Any]:
    contract_path = path or DEFAULT_CONTRACT_PATH
    payload = json.loads(contract_path.read_text(encoding="utf-8"))
    if not isinstance(payload, dict):
        raise ValueError("event_identity_contract.json must contain an object")
    weights = payload.get("weights")
    if not isinstance(weights, dict) or not math.isclose(
        sum(float(weights.get(key, 0.0)) for key in ("title", "location", "city", "source_host")),
        1.0,
        abs_tol=1e-9,
    ):
        raise ValueError("event identity weights must sum to 1.0")
    return payload


def normalize_identity_text(value: Any) -> str:
    raw = _text(value).lower()
    raw = raw.replace("ä", "ae").replace("ö", "oe").replace("ü", "ue").replace("ß", "ss")
    raw = unicodedata.normalize("NFKD", raw)
    raw = "".join(ch for ch in raw if not unicodedata.combining(ch))
    raw = re.sub(r"[^a-z0-9]+", " ", raw)
    return re.sub(r"\s+", " ", raw).strip()


def canonical_identity_url(raw: Any, contract: Mapping[str, Any] | None = None) -> str:
    value = re.sub(r"\s+", "", _text(raw))
    if not value:
        return ""
    if "://" not in value:
        value = "https://" + value
    try:
        parsed = urlparse(value)
    except Exception:
        return value.rstrip("/")
    tracking = set((contract or {}).get("tracking_parameters") or [])
    kept = []
    for key, item_value in parse_qsl(parsed.query, keep_blank_values=True):
        lower = key.lower().strip()
        if lower.startswith("utm_") or lower in tracking:
            continue
        kept.append((key, item_value))
    cleaned = parsed._replace(
        scheme=(parsed.scheme or "https").lower(),
        netloc=parsed.netloc.lower(),
        query=urlencode(kept, doseq=True),
        fragment="",
    )
    return urlunparse(cleaned).rstrip("/")


def identity_url_host(raw: Any, contract: Mapping[str, Any] | None = None) -> str:
    url = canonical_identity_url(raw, contract)
    if not url:
        return ""
    host = urlparse(url).netloc.lower()
    return host[4:] if host.startswith("www.") else host


def _value(item: Mapping[str, Any], aliases: Iterable[str]) -> str:
    for alias in aliases:
        value = _text(item.get(alias))
        if value:
            return value
    return ""


def _event_fields(item: Mapping[str, Any]) -> dict[str, str]:
    return {
        "id": _value(item, ("id_suggestion", "id", "event_id")),
        "title": _value(item, ("title", "eventName")),
        "date": _value(item, ("date", "start_date")),
        "end_date": _value(item, ("endDate", "end_date", "enddate")),
        "city": _value(item, ("city", "stadt")),
        "location": _value(item, ("location", "venue", "ort")),
        "url": _value(item, ("source_url", "url", "event_url", "official_source_url")),
    }


def _parse_day(raw: str) -> date | None:
    if not raw:
        return None
    try:
        return datetime.strptime(raw, "%Y-%m-%d").date()
    except ValueError:
        return None


def _date_ranges_overlap(a: Mapping[str, str], b: Mapping[str, str]) -> bool:
    a_start = _parse_day(a.get("date", ""))
    b_start = _parse_day(b.get("date", ""))
    if a_start is None or b_start is None:
        return False
    a_end = _parse_day(a.get("end_date", "")) or a_start
    b_end = _parse_day(b.get("end_date", "")) or b_start
    return a_start <= b_end and b_start <= a_end


def _tokens(value: str, contract: Mapping[str, Any], *, title: bool = False) -> set[str]:
    stopwords = set(contract.get("stopwords") or []) if title else set()
    aliases = dict(contract.get("token_aliases") or {})
    result: set[str] = set()
    for token in normalize_identity_text(value).split():
        token = aliases.get(token, token)
        if not token or token in stopwords:
            continue
        if title and re.fullmatch(r"20\d{2}", token):
            continue
        result.add(token)
    return result


def _token_similarity(a: str, b: str, contract: Mapping[str, Any], *, title: bool = False) -> tuple[float, int]:
    a_norm = normalize_identity_text(a)
    b_norm = normalize_identity_text(b)
    if a_norm and a_norm == b_norm:
        return 1.0, max(1, len(_tokens(a, contract, title=title)))
    a_tokens = _tokens(a, contract, title=title)
    b_tokens = _tokens(b, contract, title=title)
    if not a_tokens or not b_tokens:
        return 0.0, 0
    shared = len(a_tokens & b_tokens)
    if shared == 0:
        return 0.0, 0
    dice = (2.0 * shared) / (len(a_tokens) + len(b_tokens))
    containment = shared / min(len(a_tokens), len(b_tokens))
    score = max(dice, containment)
    if title and shared < int(contract.get("min_shared_title_tokens", 2)):
        score = min(score, float(contract.get("single_token_title_cap", 0.58)))
    return score, shared


def _score(candidate: Mapping[str, str], existing: Mapping[str, str], contract: Mapping[str, Any]) -> tuple[float, dict[str, Any]]:
    title_score, shared_title_tokens = _token_similarity(candidate["title"], existing["title"], contract, title=True)
    location_score, _ = _token_similarity(candidate["location"], existing["location"], contract)
    city_score = 1.0 if normalize_identity_text(candidate["city"]) and normalize_identity_text(candidate["city"]) == normalize_identity_text(existing["city"]) else 0.0
    candidate_host = identity_url_host(candidate["url"], contract)
    existing_host = identity_url_host(existing["url"], contract)
    source_host_score = 1.0 if candidate_host and candidate_host == existing_host else 0.0
    weights = contract["weights"]
    total = (
        float(weights["title"]) * title_score
        + float(weights["location"]) * location_score
        + float(weights["city"]) * city_score
        + float(weights["source_host"]) * source_host_score
    )
    return total, {
        "title_score": title_score,
        "shared_title_tokens": shared_title_tokens,
        "location_score": location_score,
        "city_score": city_score,
        "source_host_score": source_host_score,
    }


def _result(status: str = "none", **kwargs: Any) -> dict[str, Any]:
    return {
        "status": status,
        "matched_event_id": "",
        "matched_event_title": "",
        "matched_event_date": "",
        "matched_event_location": "",
        "matched_event_url": "",
        "score": 0.0,
        "match_type": "",
        "confidence": "",
        "reason": "",
        **kwargs,
    }


def compare_event_identity(candidate_item: Mapping[str, Any], existing_item: Mapping[str, Any], contract: Mapping[str, Any] | None = None) -> dict[str, Any]:
    cfg = dict(contract or load_event_identity_contract())
    candidate = _event_fields(candidate_item)
    existing = _event_fields(existing_item)
    same_id = bool(candidate["id"] and existing["id"] and normalize_identity_text(candidate["id"]) == normalize_identity_text(existing["id"]))
    date_overlap = _date_ranges_overlap(candidate, existing)
    if not same_id and not date_overlap:
        return _result()

    score, details = _score(candidate, existing, cfg)
    candidate_url = canonical_identity_url(candidate["url"], cfg)
    existing_url = canonical_identity_url(existing["url"], cfg)
    same_url = bool(candidate_url and candidate_url == existing_url)
    same_title = bool(normalize_identity_text(candidate["title"]) and normalize_identity_text(candidate["title"]) == normalize_identity_text(existing["title"]))
    same_location = bool(normalize_identity_text(candidate["location"]) and normalize_identity_text(candidate["location"]) == normalize_identity_text(existing["location"]))

    status = "none"
    match_type = ""
    reason = ""
    confidence = ""
    if same_id:
        if date_overlap and (same_url or (details["title_score"] >= float(cfg["review_threshold"]) and score >= float(cfg["review_threshold"]))):
            status = "same_identity"
            match_type = "same_event_id"
            reason = "Die stabile Event-ID ist bereits vorhanden und die fachlichen Merkmale passen zum selben Event."
            confidence = "high"
        else:
            status = "identity_conflict"
            match_type = "event_id_conflict"
            reason = "Die stabile Event-ID ist bereits belegt, aber Titel, Termin oder Quelle passen nicht sicher zusammen."
            confidence = "high"
    elif same_title and (same_location or details["city_score"] == 1.0):
        status = "exact"
        match_type = "same_title_and_date"
        reason = "Titel und Termin stimmen mit einem vorhandenen Event überein."
        confidence = "high"
        score = max(score, 0.99)
    elif same_url and details["shared_title_tokens"] >= 1:
        status = "possible"
        match_type = "same_source_url_and_date"
        reason = "Dieselbe kanonische Quelle und derselbe Termin sind bereits vorhanden; die abweichende Bezeichnung muss fachlich geprüft werden."
        confidence = "high"
        score = max(score, float(cfg["review_threshold"]))
    elif score >= float(cfg["review_threshold"]) and details["shared_title_tokens"] >= int(cfg.get("min_shared_title_tokens", 2)):
        status = "possible"
        match_type = "semantic_title_date_context"
        reason = "Termin und prägende Titelbegriffe stimmen stark überein; Ort, Stadt oder Quellkontext stützen den Verdacht."
        confidence = "high" if score >= 0.86 else "medium"

    if status == "none":
        return _result()
    return _result(
        status,
        matched_event_id=existing["id"],
        matched_event_title=existing["title"],
        matched_event_date=existing["date"],
        matched_event_location=existing["location"],
        matched_event_url=existing_url,
        score=round(min(1.0, max(0.0, score)), 3),
        match_type=match_type,
        confidence=confidence,
        reason=reason,
        details=details,
    )


def event_rows_from_sheet_values(values: Iterable[Iterable[Any]]) -> list[dict[str, str]]:
    rows = [list(row) for row in values if isinstance(row, (list, tuple))]
    header_index = -1
    header: list[str] = []
    for offset, raw_header in enumerate(rows):
        normalized = [normalize_identity_text(value).replace(" ", "_") for value in raw_header]
        if "title" not in normalized or "date" not in normalized:
            continue
        if not ({"id", "event_id"} & set(normalized)):
            continue
        header_index = offset
        header = normalized
        break
    if header_index < 0:
        raise ValueError("event sheet header not found")
    result: list[dict[str, str]] = []
    for raw in rows[header_index + 1 :]:
        if not any(_text(value) for value in raw):
            continue
        item = {name: _text(raw[pos]) if pos < len(raw) else "" for pos, name in enumerate(header) if name}
        if not item.get("id") and not item.get("event_id"):
            raise ValueError("event row without stable id")
        if not item.get("title") or not item.get("date"):
            raise ValueError("event row without title or date")
        result.append(item)
    return result


def find_best_event_match(candidate: Mapping[str, Any], existing_events: Iterable[Mapping[str, Any]], contract: Mapping[str, Any] | None = None) -> dict[str, Any]:
    cfg = dict(contract or load_event_identity_contract())
    rank = {"none": 0, "possible": 1, "same_identity": 2, "exact": 3, "identity_conflict": 4}
    best = _result()
    for existing in existing_events:
        if not isinstance(existing, Mapping):
            continue
        result = compare_event_identity(candidate, existing, cfg)
        current_key = (rank.get(result["status"], 0), float(result.get("score", 0.0)), result.get("matched_event_id", ""))
        best_key = (rank.get(best["status"], 0), float(best.get("score", 0.0)), best.get("matched_event_id", ""))
        if current_key > best_key:
            best = result
    return best


def apply_event_identity_match(candidate: Mapping[str, Any], match: Mapping[str, Any], *, allow_same_identity: bool = False) -> dict[str, Any]:
    enriched = dict(candidate)
    status = _text(match.get("status"))
    if status == "none" or (status == "same_identity" and allow_same_identity):
        for field in (
            "matched_event_id",
            "match_score",
            "duplicate_score",
            "duplicate_confidence",
            "duplicate_reason",
            "duplicate_match_type",
            "matched_event_title",
            "matched_event_date",
            "matched_event_location",
            "matched_event_url",
        ):
            enriched[field] = ""
        if normalize_identity_text(enriched.get("duplicate_status", "")) == "review":
            enriched["duplicate_status"] = ""
        enriched["hard_duplicate"] = False
        enriched["event_identity_status"] = status or "none"
        return enriched

    matched_id = _text(match.get("matched_event_id"))
    old_status = normalize_identity_text(enriched.get("duplicate_status", ""))
    old_match = _text(enriched.get("matched_event_id", ""))
    human_distinct = old_status == "distinct" and old_match and old_match == matched_id

    enriched.update(
        {
            "matched_event_id": matched_id,
            "match_score": f"{float(match.get('score', 0.0)):.3f}",
            "duplicate_score": f"{float(match.get('score', 0.0)):.3f}",
            "duplicate_confidence": _text(match.get("confidence")),
            "duplicate_reason": _text(match.get("reason")),
            "duplicate_match_type": _text(match.get("match_type")),
            "matched_event_title": _text(match.get("matched_event_title")),
            "matched_event_date": _text(match.get("matched_event_date")),
            "matched_event_location": _text(match.get("matched_event_location")),
            "matched_event_url": _text(match.get("matched_event_url")),
            "event_identity_status": status,
            "hard_duplicate": not human_distinct,
        }
    )
    if not human_distinct:
        enriched["duplicate_status"] = "review"
    return enriched

File: api/control-center/_event_identity.php
<?php
declare(strict_types=1);

/**
 * Canonical, side-effect-free event identity contract. The same JSON weights,
 * stopwords and thresholds are consumed by the Python intake matcher.
 */
function be_cc_event_identity_contract(): array
{
    static $contract = null;
    if (is_array($contract)) return $contract;
    $path = dirname(__DIR__, 2) . '/data/event_identity_contract.json';
    if (!is_file($path)) throw new RuntimeException('Der Event-Identitätsvertrag fehlt.');
    $decoded = json_decode((string)file_get_contents($path), true);
    if (!is_array($decoded)) throw new RuntimeException('Der Event-Identitätsvertrag ist ungültig.');
    $weights = is_array($decoded['weights'] ?? null) ? $decoded['weights'] : [];
    $sum = 0.0;
    foreach (['title','location','city','source_host'] as $key) $sum += (float)($weights[$key] ?? 0.0);
    if (abs($sum - 1.0) > 0.000000001) throw new RuntimeException('Die Event-Identitätsgewichte müssen 1 ergeben.');
    return $contract = $decoded;
}

function be_cc_event_identity_text(mixed $value): string
{
    $raw = strtr(trim((string)$value), ['Ä'=>'Ae','Ö'=>'Oe','Ü'=>'Ue','ä'=>'ae','ö'=>'oe','ü'=>'ue','ß'=>'ss']);
    $ascii = function_exists('iconv') ? @iconv('UTF-8', 'ASCII//TRANSLIT//IGNORE', $raw) : false;
    $raw = strtolower(is_string($ascii) && $ascii !== '' ? $ascii : $raw);
    $raw = (string)preg_replace('/[^a-z0-9]+/u', ' ', $raw);
    return trim((string)preg_replace('/\s+/u', ' ', $raw));
}

function be_cc_event_identity_value(array $item, array $aliases): string
{
    foreach ($aliases as $alias) {
        $value = trim((string)($item[$alias] ?? ''));
        if ($value !== '') return $value;
    }
    return '';
}

function be_cc_event_identity_fields(array $item): array
{
    return [
        'id'=>be_cc_event_identity_value($item, ['id_suggestion','id','event_id']),
        'title'=>be_cc_event_identity_value($item, ['title','eventName']),
        'date'=>be_cc_event_identity_value($item, ['date','start_date']),
        'end_date'=>be_cc_event_identity_value($item, ['endDate','end_date','enddate']),
        'city'=>be_cc_event_identity_value($item, ['city','stadt']),
        'location'=>be_cc_event_identity_value($item, ['location','venue','ort']),
        'url'=>be_cc_event_identity_value($item, ['source_url','url','event_url','official_source_url']),
    ];
}

function be_cc_event_identity_url(mixed $raw, ?array $contract = null): string
{
    $value = (string)preg_replace('/\s+/u', '', trim((string)$raw));
    if ($value === '') return '';
    if (!str_contains($value, '://')) $value = 'https://' . $value;
    $parts = parse_url($value);
    if (!is_array($parts)) return rtrim($value, '/');
    $scheme = strtolower((string)($parts['scheme'] ?? 'https'));
    $host = strtolower((string)($parts['host'] ?? ''));
    if ($host === '') return rtrim($value, '/');
    $port = isset($parts['port']) ? ':' . (string)$parts['port'] : '';
    $path = (string)($parts['path'] ?? '');
    $query = [];
    parse_str((string)($parts['query'] ?? ''), $query);
    $tracking = array_fill_keys(array_map('strval', (array)(($contract ?? be_cc_event_identity_contract())['tracking_parameters'] ?? [])), true);
    foreach (array_keys($query) as $key) {
        $lower = strtolower(trim((string)$key));
        if (str_starts_with($lower, 'utm_') || isset($tracking[$lower])) unset($query[$key]);
    }
    $queryText = $query ? '?' . http_build_query($query, '', '&', PHP_QUERY_RFC3986) : '';
    return rtrim($scheme . '://' . $host . $port . $path . $queryText, '/');
}

function be_cc_event_identity_host(mixed $raw, ?array $contract = null): string
{
    $url = be_cc_event_identity_url($raw, $contract);
    if ($url === '') return '';
    $host = strtolower((string)(parse_url($url, PHP_URL_HOST) ?? ''));
    return str_starts_with($host, 'www.') ? substr($host, 4) : $host;
}

function be_cc_event_identity_day(string $raw): ?DateTimeImmutable
{
    if ($raw === '') return null;
    $day = DateTimeImmutable::createFromFormat('!Y-m-d', $raw);
    $errors = DateTimeImmutable::getLastErrors();
    if (!$day || (is_array($errors) && ((int)$errors['warning_count'] > 0 || (int)$errors['error_count'] > 0))) return null;
    return $day->format('Y-m-d') === $raw ? $day : null;
}

function be_cc_event_identity_ranges_overlap(array $a, array $b): bool
{
    $aStart = be_cc_event_identity_day((string)($a['date'] ?? ''));
    $bStart = be_cc_event_identity_day((string)($b['date'] ?? ''));
    if (!$aStart || !$bStart) return false;
    $aEnd = be_cc_event_identity_day((string)($a['end_date'] ?? '')) ?: $aStart;
    $bEnd = be_cc_event_identity_day((string)($b['end_date'] ?? '')) ?: $bStart;
    return $aStart <= $bEnd && $bStart <= $aEnd;
}

function be_cc_event_identity_tokens(string $value, array $contract, bool $title = false): array
{
    $stopwords = $title ? array_fill_keys(array_map('strval', (array)($contract['stopwords'] ?? [])), true) : [];
    $aliases = is_array($contract['token_aliases'] ?? null) ? $contract['token_aliases'] : [];
    $tokens = [];
    foreach (explode(' ', be_cc_event_identity_text($value)) as $token) {
        $token = trim((string)($aliases[$token] ?? $token));
        if ($token === '' || isset($stopwords[$token])) continue;
        if ($title && preg_match('/^20\d{2}$/', $token) === 1) continue;
        $tokens[$token] = true;
    }
    return array_keys($tokens);
}

function be_cc_event_identity_token_similarity(string $a, string $b, array $contract, bool $title = false): array
{
    $aNorm = be_cc_event_identity_text($a);
    $bNorm = be_cc_event_identity_text($b);
    if ($aNorm !== '' && $aNorm === $bNorm) {
        return ['score'=>1.0, 'shared'=>max(1, count(be_cc_event_identity_tokens($a, $contract, $title)))];
    }
    $aTokens = be_cc_event_identity_tokens($a, $contract, $title);
    $bTokens = be_cc_event_identity_tokens($b, $contract, $title);
    if (!$aTokens || !$bTokens) return ['score'=>0.0, 'shared'=>0];
    $shared = count(array_intersect($aTokens, $bTokens));
    if ($shared < 1) return ['score'=>0.0, 'shared'=>0];
    $dice = (2.0 * $shared) / (count($aTokens) + count($bTokens));
    $containment = $shared / min(count($aTokens), count($bTokens));
    $score = max($dice, $containment);
    if ($title && $shared < (int)($contract['min_shared_title_tokens'] ?? 2)) {
        $score = min($score, (float)($contract['single_token_title_cap'] ?? 0.58));
    }
    return ['score'=>$score, 'shared'=>$shared];
}

function be_cc_event_identity_score(array $candidate, array $existing, array $contract): array
{
    $title = be_cc_event_identity_token_similarity((string)$candidate['title'], (string)$existing['title'], $contract, true);
    $location = be_cc_event_identity_token_similarity((string)$candidate['location'], (string)$existing['location'], $contract, false);
    $candidateCity = be_cc_event_identity_text((string)$candidate['city']);
    $existingCity = be_cc_event_identity_text((string)$existing['city']);
    $cityScore = $candidateCity !== '' && $candidateCity === $existingCity ? 1.0 : 0.0;
    $candidateHost = be_cc_event_identity_host((string)$candidate['url'], $contract);
    $existingHost = be_cc_event_identity_host((string)$existing['url'], $contract);
    $hostScore = $candidateHost !== '' && $candidateHost === $existingHost ? 1.0 : 0.0;
    $weights = (array)$contract['weights'];
    $score = (float)$weights['title'] * (float)$title['score']
        + (float)$weights['location'] * (float)$location['score']
        + (float)$weights['city'] * $cityScore
        + (float)$weights['source_host'] * $hostScore;
    return [
        'score'=>$score,
        'title_score'=>(float)$title['score'],
        'shared_title_tokens'=>(int)$title['shared'],
        'location_score'=>(float)$location['score'],
        'city_score'=>$cityScore,
        'source_host_score'=>$hostScore,
    ];
}

function be_cc_event_identity_empty_result(): array
{
    return [
        'status'=>'none','matched_event_id'=>'','matched_event_title'=>'','matched_event_date'=>'',
        'matched_event_location'=>'','matched_event_url'=>'','score'=>0.0,'match_type'=>'',
        'confidence'=>'','reason'=>'',
    ];
}

function be_cc_event_identity_compare(array $candidateItem, array $existingItem, ?array $contract = null): array
{
    $cfg = $contract ?? be_cc_event_identity_contract();
    $candidate = be_cc_event_identity_fields($candidateItem);
    $existing = be_cc_event_identity_fields($existingItem);
    $sameId = $candidate['id'] !== '' && $existing['id'] !== '' && be_cc_event_identity_text($candidate['id']) === be_cc_event_identity_text($existing['id']);
    $dateOverlap = be_cc_event_identity_ranges_overlap($candidate, $existing);
    if (!$sameId && !$dateOverlap) return be_cc_event_identity_empty_result();

    $details = be_cc_event_identity_score($candidate, $existing, $cfg);
    $candidateUrl = be_cc_event_identity_url($candidate['url'], $cfg);
    $existingUrl = be_cc_event_identity_url($existing['url'], $cfg);
    $sameUrl = $candidateUrl !== '' && $candidateUrl === $existingUrl;
    $sameTitle = be_cc_event_identity_text($candidate['title']) !== '' && be_cc_event_identity_text($candidate['title']) === be_cc_event_identity_text($existing['title']);
    $sameLocation = be_cc_event_identity_text($candidate['location']) !== '' && be_cc_event_identity_text($candidate['location']) === be_cc_event_identity_text($existing['location']);

    $status = 'none';
    $matchType = '';
    $reason = '';
    $confidence = '';
    $score = (float)$details['score'];
    if ($sameId) {
        if ($dateOverlap && ($sameUrl || ((float)$details['title_score'] >= (float)$cfg['review_threshold'] && $score >= (float)$cfg['review_threshold']))) {
            $status = 'same_identity';
            $matchType = 'same_event_id';
            $reason = 'Die stabile Event-ID ist bereits vorhanden und die fachlichen Merkmale passen zum selben Event.';
            $confidence = 'high';
        } else {
            $status = 'identity_conflict';
            $matchType = 'event_id_conflict';
            $reason = 'Die stabile Event-ID ist bereits belegt, aber Titel, Termin oder Quelle passen nicht sicher zusammen.';
            $confidence = 'high';
        }
    } elseif ($sameTitle && ($sameLocation || $sameUrl)) {
        $status = 'exact';
        $matchType = 'same_title_and_date';
        $reason = 'Titel und Termin stimmen mit einem vorhandenen Event überein.';
        $confidence = 'high';
        $score = max($score, 0.99);
    } elseif ($sameUrl && (int)$details['shared_title_tokens'] >= 1) {
        $status = 'possible';
        $matchType = 'same_source_url_and_date';
        $reason = 'Dieselbe kanonische Quelle und derselbe Termin sind bereits vorhanden; die abweichende Bezeichnung muss fachlich geprüft werden.';
        $confidence = 'high';
        $score = max($score, (float)$cfg['review_threshold']);
    } elseif ($score >= (float)$cfg['review_threshold'] && (int)$details['shared_title_tokens'] >= (int)($cfg['min_shared_title_tokens'] ?? 2)) {
        $status = 'possible';
        $matchType = 'semantic_title_date_context';
        $reason = 'Termin und prägende Titelbegriffe stimmen stark überein; Ort, Stadt oder Quellkontext stützen den Verdacht.';
        $confidence = $score >= 0.86 ? 'high' : 'medium';
    }

    if ($status === 'none') return be_cc_event_identity_empty_result();
    return [
        'status'=>$status,
        'matched_event_id'=>$existing['id'],
        'matched_event_title'=>$existing['title'],
        'matched_event_date'=>$existing['date'],
        'matched_event_location'=>$existing['location'],
        'matched_event_url'=>$existingUrl,
        'score'=>round(min(1.0, max(0.0, $score)), 3),
        'match_type'=>$matchType,
        'confidence'=>$confidence,
        'reason'=>$reason,
        'details'=>$details,
    ];
}

function be_cc_event_identity_find_best(array $candidate, array $events, ?array $contract = null): array
{
    $cfg = $contract ?? be_cc_event_identity_contract();
    $ranks = ['none'=>0,'possible'=>1,'same_identity'=>2,'exact'=>3,'identity_conflict'=>4];
    $best = be_cc_event_identity_empty_result();
    foreach ($events as $existing) {
        if (!is_array($existing)) continue;
        $result = be_cc_event_identity_compare($candidate, $existing, $cfg);
        $current = [$ranks[$result['status']] ?? 0, (float)($result['score'] ?? 0.0), (string)($result['matched_event_id'] ?? '')];
        $previous = [$ranks[$best['status']] ?? 0, (float)($best['score'] ?? 0.0), (string)($best['matched_event_id'] ?? '')];
        if ($current > $previous) $best = $result;
    }
    return $best;
}

function be_cc_event_identity_enrich(array $candidate, array $events, bool $allowSameIdentity = true): array
{
    $match = be_cc_event_identity_find_best($candidate, $events);
    $status = trim((string)($match['status'] ?? 'none'));
    if ($status === 'none' || ($status === 'same_identity' && $allowSameIdentity)) {
        foreach ([
            'matched_event_id','match_score','duplicate_score','duplicate_confidence','duplicate_reason',
            'duplicate_match_type','matched_event_title','matched_event_date','matched_event_location','matched_event_url',
        ] as $field) $candidate[$field] = '';
        if (be_cc_event_identity_text($candidate['duplicate_status'] ?? '') === 'review') $candidate['duplicate_status'] = '';
        $candidate['hard_duplicate'] = false;
        $candidate['event_identity_status'] = $status;
        return $candidate;
    }

    $matchedId = trim((string)($match['matched_event_id'] ?? ''));
    $oldStatus = be_cc_event_identity_text($candidate['duplicate_status'] ?? '');
    $oldMatch = trim((string)($candidate['matched_event_id'] ?? ''));
    $humanDistinct = $oldStatus === 'distinct' && $oldMatch !== '' && $oldMatch === $matchedId;
    $score = number_format((float)($match['score'] ?? 0.0), 3, '.', '');
    $candidate = array_replace($candidate, [
        'matched_event_id'=>$matchedId,
        'match_score'=>$score,
        'duplicate_score'=>$score,
        'duplicate_confidence'=>(string)($match['confidence'] ?? ''),
        'duplicate_reason'=>(string)($match['reason'] ?? ''),
        'duplicate_match_type'=>(string)($match['match_type'] ?? ''),
        'matched_event_title'=>(string)($match['matched_event_title'] ?? ''),
        'matched_event_date'=>(string)($match['matched_event_date'] ?? ''),
        'matched_event_location'=>(string)($match['matched_event_location'] ?? ''),
        'matched_event_url'=>(string)($match['matched_event_url'] ?? ''),
        'event_identity_status'=>$status,
        'hard_duplicate'=>!$humanDistinct,
    ]);
    if (!$humanDistinct) $candidate['duplicate_status'] = 'review';
    return $candidate;
}

function be_cc_event_identity_table_rows(string $tab): array
{
    $response = be_google_sheets_values_get($tab . '!A:AZ');
    $values = is_array($response['values'] ?? null) ? $response['values'] : [];
    $headerOffset = -1;
    $index = [];
    foreach ($values as $offset=>$row) {
        if (!is_array($row)) continue;
        $normalized = array_map(static fn($value): string => be_cc_event_identity_text($value), $row);
        if (!in_array('title', $normalized, true) || !in_array('date', $normalized, true)) continue;
        if (!in_array('id', $normalized, true) && !in_array('event id', $normalized, true) && !in_array('event_id', $normalized, true)) continue;
        $headerOffset = (int)$offset;
        foreach ($normalized as $position=>$name) if ($name !== '') $index[$name] = (int)$position;
        break;
    }
    if ($headerOffset < 0) throw new RuntimeException('Der Event-Header wurde nicht gefunden: ' . $tab);
    $rows = [];
    for ($i=$headerOffset+1;$i<count($values);$i++) {
        $raw = is_array($values[$i] ?? null) ? $values[$i] : [];
        if (!array_filter($raw, static fn($value): bool => trim((string)$value) !== '')) continue;
        $row = [];
        foreach ($index as $name=>$position) $row[$name] = trim((string)($raw[$position] ?? ''));
        if (isset($row['event id']) && !isset($row['event_id'])) $row['event_id'] = $row['event id'];
        $fields = be_cc_event_identity_fields($row);
        if ($fields['id'] === '') throw new RuntimeException('Der Eventbestand enthält eine Zeile ohne stabile ID: ' . $tab . '.');
        if ($fields['title'] === '' || $fields['date'] === '') throw new RuntimeException('Der Eventbestand enthält eine Zeile ohne Titel oder Datum: ' . $tab . '.');
        $rows[] = $row;
    }
    return $rows;
}

function be_cc_event_identity_merge_event_rows(array $baseRows, array $overlayRows): array
{
    $byKey = [];
    $order = [];
    $overlayKeys = [];
    $add = static function(array $row, bool $overlay) use (&$byKey, &$order, &$overlayKeys): void {
        $fields = be_cc_event_identity_fields($row);
        $id = be_cc_event_identity_text($fields['id']);
        if ($id === '') throw new RuntimeException('Der Eventbestand enthält eine Zeile ohne stabile ID.');
        $key = 'id:' . $id;
        if (!isset($byKey[$key])) {
            $order[] = $key;
        } elseif (!$overlay || isset($overlayKeys[$key])) {
            throw new RuntimeException('Der Eventbestand enthält eine doppelte stabile ID.');
        }
        if ($overlay) $overlayKeys[$key] = true;
        $byKey[$key] = $row;
    };
    foreach ($baseRows as $row) if (is_array($row)) $add($row, false);
    foreach ($overlayRows as $row) if (is_array($row)) $add($row, true);
    return array_map(static fn(string $key): array => $byKey[$key], $order);
}

function be_cc_event_identity_current_events(): array
{
    static $cache = [];
    $environment = strtolower(trim(be_app_env_value()));
    if (isset($cache[$environment])) return $cache[$environment];
    $base = be_cc_event_identity_table_rows('Events');
    if (in_array($environment, ['staging','development','dev','local','test'], true)) {
        $overlay = be_cc_event_identity_table_rows('Events_Staging');
        return $cache[$environment] = be_cc_event_identity_merge_event_rows($base, $overlay);
    }
    if (in_array($environment, ['live','production','prod'], true)) return $cache[$environment] = $base;
    throw new RuntimeException('Der Eventbestand kann für diese Laufzeitumgebung nicht sicher bestimmt werden.');
}

function be_cc_event_identity_enrich_current(array $candidate, bool $allowSameIdentity = true): array
{
    return be_cc_event_identity_enrich($candidate, be_cc_event_identity_current_events(), $allowSameIdentity);
}
